### PR TITLE
[autodiff] Support passing requires_grad=True tensor to an arg with needs_grad=False

### PR DESCRIPTION
**Pull request by ailzhang**
Issue: #

### <samp>🤖 Generated by Copilot at a3e51bb</samp>

This pull request adds support for the `needs_grad` attribute of the `NdarrayType` annotation, which allows specifying whether an ndarray argument requires gradient computation or not. It modifies the `extract_arg` and `check_matched` functions in `kernel_impl.py` and `ndarray_type.py` to handle the attribute, and adds two test cases in `test_ad_ndarray.py` to verify the behavior.

### <samp>🤖 Generated by Copilot at a3e51bb</samp>

*  Add logic to handle `needs_grad` attribute of `NdarrayType` annotation in `extract_arg` function ([link](https://github.com/taichi-dev/taichi/pull/7970/files?diff=unified&w=0#diff-a157043b38542c8145447ff342fda65fe4d54fb777fe514daa70007e83e20dc1L392-R436))
*  Relax condition for matching `needs_grad` attribute of `NdarrayType` annotation and actual ndarray type in `check_matched` function ([link](https://github.com/taichi-dev/taichi/pull/7970/files?diff=unified&w=0#diff-06e0109e9fa5071f7d364306981845d410fa17425db48001e3ba69337b47c152L131-R132))
*  Add test cases to verify `needs_grad` attribute of `NdarrayType` annotation in `test_ad_ndarray.py` ([link](https://github.com/taichi-dev/taichi/pull/7970/files?diff=unified&w=0#diff-015ebf0e792298ff88ac9c4d87efba82414b3a19a4d85d0887d16099dfe2fc09R1371-R1417))

### CODE REVIEW

**Review by erizmr — APPROVED** [review]
LGTM!
